pkg/rules/rules.go

```diff
@@ -86,15 +86,21 @@ func filterRules(ruleGroups []*rulespb.RuleGroup, matcherSets [][]*labels.Matche
 		return ruleGroups
 	}
 
-	for _, g := range ruleGroups {
-		filteredRules := g.Rules[:0]
-		for _, r := range g.Rules {
+	for i := 0; i < len(ruleGroups); i++ {
+		filteredRules := ruleGroups[i].Rules[:0]
+		for _, r := range ruleGroups[i].Rules {
 			rl := r.GetLabels()
 			if matches(matcherSets, rl) {
 				filteredRules = append(filteredRules, r)
 			}
 		}
-		g.Rules = filteredRules
+
+		ruleGroups[i].Rules = filteredRules
+		// Filter Groups which do not have any rules.
+		if len(ruleGroups[i].Rules) == 0 {
+			ruleGroups = append(ruleGroups[:i], ruleGroups[i+1:]...)
+			i-- // -1 as the slice just got shorter
+		}
```

Comment:
This double `for` loop is becoming complicated (modifying the loop variable `i` inside the loop itself is a bad sign) because it wants to modify `ruleGroups` in-place. Why not to create a kind of "result" slice and you only append to it if there are still rules present in a group after the filtering?

Reply:
Yeah, it's a bit confusing, so using https://github.com/golang/go/wiki/SliceTricks#filter-in-place instead. That way it's not violating best practice + not allocating new variables, here:  https://github.com/thanos-io/thanos/pull/5537/commits/18a69bb837fdb19c3b102d2ca389cef3122894f6 🙂